pageserver/ctl/src/layers.rs

```diff
@@ -70,11 +79,47 @@ fn read_delta_file(path: impl AsRef<Path>) -> Result<()> {
         },
     )?;
     let mut cursor = BlockCursor::new(&file);
+
+    let mut result = vec![];
     for (k, v) in all {
         let value = cursor.read_blob(v.pos())?;
-        println!("key:{} value_len:{}", k, value.len());
+        result.push((k, value.len()));
     }
     // TODO(chi): special handling for last key?
+    Ok(result)
+}
+
+// We divide the entire i128 keyspace into pre-assigned fixed segments,
+// 8MB each. Group keys by segment, and report segment size for each.
+//
+// 8MB is chosen as the segment size because we're unlikely to make
+// s3 partial downloads smaller than 8KB (due to cost). So summarizing
+// layer metadata in 8KB segments could be enough to generate good test
+// data for write amplification tests.
+fn read_delta_segments(path: impl AsRef<Path>) -> Result<Vec<(i128, usize)>> {
+    fn key_to_segment(key: &Key) -> i128 {
+        key.to_i128() >> 26
+    }
+
+    use itertools::Itertools;
+    let delta_metadata = read_delta_file(path)?;
+    let group_iter = delta_metadata.iter().group_by(|(k, _)| key_to_segment(k));
+    let group_sizes = group_iter
+        .into_iter()
+        .map(|(segment, lengths_group)| {
+            let lengths: Vec<_> = lengths_group.collect();
+            let sum: usize = lengths.iter().map(|(_k, len)| len).sum();
+            (segment, sum)
```

Comment:
```suggestion
            let sum: usize = lengths_group.map(|(_k, len)| len).sum();
            (segment, sum)
```

The lengths is actually a `Vec<(key, len)>` so this should be much faster.